File: php/IpSearch.php
<?php
/**
 * 使用PHP代码从[qqzeng-ip]数据库的二进制文件中获取IP地址所在地理位置信息
 * @qqzeng-ip
 */
 define('filepath' ,"ip.dat");
class IpLocation {
	 var $StartIP = 0;
    var $EndIP = 0;
    var $Country = '';
    var $Local = '';
 
    var $CountryFlag = 0; // 标识 Country位置
        // 0x01,随后3字节为Country偏移,没有Local
        // 0x02,随后3字节为Country偏移,接着是Local
        // 其他,Country,Local,Local有类似的压缩。可能多重引用。
 
    var $fp;
 
    var $FirstStartIp = 0;
    var $LastStartIp = 0;
    var $EndIpOff = 0;
 
    function getStartIp($RecNo)
    {
        $offset = $this->FirstStartIp + $RecNo * 7;
        @fseek($this->fp, $offset, SEEK_SET);
        $buf = fread($this->fp, 7);
        $this->EndIpOff = ord($buf[4]) + (ord($buf[5]) * 256) + (ord($buf[6]) * 256 * 256);
        $this->StartIp = ord($buf[0]) + (ord($buf[1]) * 256) + (ord($buf[2]) * 256 * 256) + (ord($buf[3]) * 256 * 256 * 256);
        return $this->StartIp;
    }
 
    function getEndIp()
    {
        @fseek($this->fp, $this->EndIpOff, SEEK_SET);
        $buf = fread($this->fp, 5);
        $this->EndIp = ord($buf[0]) + (ord($buf[1]) * 256) + (ord($buf[2]) * 256 * 256) + (ord($buf[3]) * 256 * 256 * 256);
        $this->CountryFlag = ord($buf[4]);
        return $this->EndIp;
    }
 
    function getCountry()
    {
        switch($this->CountryFlag)
        {
            case 1: 
            case 2: 
               
            default: 
                $this->Country = $this->getFlagStr($this->EndIpOff + 4); 
				 $this->Local =(1 == $this->CountryFlag) ?  $this->getFlagStr(ftell($this->fp)) :$this->getFlagStr($this->EndIpOff + 8); 
               
        }
    }
 
    function getFlagStr($offset)
    {
        $flag = 0;
         
        while(1)
        {
            @fseek($this->fp, $offset, SEEK_SET);
            $flag = ord(fgetc($this->fp));
             
            if($flag == 1 || $flag == 2)
            {
                $buf = fread($this->fp, 3);
                 
                if($flag == 2)
                {
                    $this->CountryFlag = 2;
                    $this->EndIpOff = $offset - 4;
                }
                 
                $offset = ord($buf[0]) + (ord($buf[1]) * 256) + (ord($buf[2]) * 256 * 256);
            }
            else
                break;
        }
         
        if($offset < 12) return '';
         
        @fseek($this->fp, $offset, SEEK_SET);
 
        return $this->getStr();
    }
 
    function getStr()
    {
        $str = '';
         
        while(1)
        {
            $c = fgetc($this->fp);
 
            if(ord($c[0]) == 0) break;
             
            $str .= $c;
        }
         
        return $str;
    }
 
    function seek($ip_address = '')
    {
        if(!$ip_address) return;
         
       
 
        $nRet;
        $ip = $this->IpToInt($ip_address);
        $this->fp = fopen(filepath, "rb");
         
        if($this->fp == NULL)
        {
            $szLocal = "OpenFileError";
            return 1;
        }
         
        @fseek($this->fp, 0, SEEK_SET);
        $buf = fread($this->fp, 8);
        $this->FirstStartIp = ord($buf[0]) + (ord($buf[1]) * 256) + (ord($buf[2]) * 256 * 256) + (ord($buf[3]) * 256 * 256 * 256);
        $this->LastStartIp = ord($buf[4]) + (ord($buf[5]) * 256) + (ord($buf[6]) * 256 * 256) + (ord($buf[7]) * 256 * 256 * 256);
 
        $RecordCount = floor(($this->LastStartIp - $this->FirstStartIp) / 7);
         
        if($RecordCount <= 1)
        {
            $this->Country = "";
            fclose($this->fp) ;
            return 2 ;
        }
 
        $RangB = 0;
        $RangE = $RecordCount;
         
        // Match ...
        while($RangB < $RangE - 1)
        {
            $RecNo = floor(($RangB + $RangE) / 2);
            $this->getStartIp($RecNo) ;
 
            if($ip == $this->StartIp)
            {
                $RangB = $RecNo;
                break;
            }
             
            if($ip > $this->StartIp) $RangB = $RecNo;
            else $RangE = $RecNo;
        }
         
        $this->getStartIp($RangB);
        $this->getEndIp();
 
        if(($this->StartIp <= $ip) && ($this->EndIp >= $ip))
        {
            $this->getCountry();
        }
        else
        {
            $this->Country = '无';
            $this->Local = '';
        }
         
        fclose($this->fp);
    }
 
    function IpToInt($Ip)
    {
        $array = explode('.', $Ip);
        $Int = ($array[0] * 256 * 256 * 256) + ($array[1] * 256 * 256) + ($array[2] * 256) + $array[3];
 
        return $Int;
    }
}

File: c#/IPSearch.cs
using System;
using System.Text;
using System.IO;
using System.Net;
using System.Net.Sockets;

namespace qqzengIP
{
    public class IPEntity
    {
        public string IP { get; set; }
        public string Country { get; set; }
        public string Local { get; set; }
    }

    public class IPSearch
    {
        static Encoding encoding = Encoding.GetEncoding("GB2312");

        private byte[] data;
        int firstStartIpOffset;
        int lastStartIpOffset;
        int ipCount;

        public int Count { get { return ipCount; } }

        public IPSearch(string dataPath)
        {
            using (FileStream fs = new FileStream(dataPath, FileMode.Open, FileAccess.Read, FileShare.Read))
            {
                data = new byte[fs.Length];
                fs.Read(data, 0, data.Length);
            }

            firstStartIpOffset = (int)data[0] + (((int)data[1]) << 8) + (((int)data[2]) << 16) + (((int)data[3]) << 24);
            lastStartIpOffset = (int)data[4] + (((int)data[5]) << 8) + (((int)data[6]) << 16) + (((int)data[7]) << 24);
            ipCount = (lastStartIpOffset - firstStartIpOffset) / 7 + 1;

            if (ipCount <= 1)
            {
                throw new ArgumentException("文件错误");
            }
        }

        public static uint IpToInt(string ip)
        {
            byte[] bytes = IPAddress.Parse(ip).GetAddressBytes();
            return (uint)bytes[3] + (((uint)bytes[2]) << 8) + (((uint)bytes[1]) << 16) + (((uint)bytes[0]) << 24);
        }

        public static string IntToIP(uint ip_Int)
        {
            return new IPAddress(ip_Int).ToString();
        }

        public IPEntity Query(string ip)
        {
            IPAddress address = IPAddress.Parse(ip);
            if (address.AddressFamily != AddressFamily.InterNetwork)
            {
                throw new ArgumentException("不支持非IPV4的地址");
            }
            if (IPAddress.IsLoopback(address))
            {
                return new IPEntity() { IP = ip, Country = "本机内部环回地址", Local = string.Empty };
            }
            long intIP = IpToInt(ip);


            IPEntity ipLocation = new IPEntity() { IP = ip };

            uint right = (uint)ipCount;
            uint left = 0;
            uint middle = 0;
            uint startIp = 0;
            uint endIpOff = 0;
            uint endIp = 0;
            int countryFlag = 0;

            while (left < (right - 1))
            {
                middle = (right + left) / 2;
                startIp = GetStartIp(middle, out endIpOff);
                if (intIP == startIp)
                {
                    left = middle;
                    break;
                }
                if (intIP > startIp)
                {
                    left = middle;
                }
                else
                {
                    right = middle;
                }
            }
            startIp = GetStartIp(left, out endIpOff);
            endIp = GetEndIp(endIpOff, out countryFlag);
            if ((startIp <= intIP) && (endIp >= intIP))
            {
                string local;
                ipLocation.Country = GetCountry(endIpOff, countryFlag, out local);
                ipLocation.Local = local;
            }
            else
            {
                ipLocation.Country = "未知";
                ipLocation.Local = string.Empty;
            }
            return ipLocation;
        }

        private uint GetStartIp(uint left, out uint endIpOff)
        {
            int leftOffset = (int)(firstStartIpOffset + (left * 7));
            endIpOff = (uint)data[4 + leftOffset] + (((uint)data[5 + leftOffset]) << 8) + (((uint)data[6 + leftOffset]) << 16);
            return (uint)data[leftOffset] + (((uint)data[1 + leftOffset]) << 8) + (((uint)data[2 + leftOffset]) << 16) + (((uint)data[3 + leftOffset]) << 24);
        }

        private uint GetEndIp(uint endIpOff, out int countryFlag)
        {
            countryFlag = data[4 + endIpOff];
            return (uint)data[endIpOff] + (((uint)data[1 + endIpOff]) << 8) + (((uint)data[2 + endIpOff]) << 16) + (((uint)data[3 + endIpOff]) << 24);
        }

        /// <summary>
        /// 获取国家
        /// </summary>
        /// <param name="endIpOff">结束ip位置</param>
        /// <param name="countryFlag">模式</param>
        /// <param name="local">返回 省市区 运营商</param>
        /// <returns>国家</returns>
        private string GetCountry(uint endIpOff, int countryFlag, out string local)
        {
            string country = string.Empty;
            uint offset = endIpOff + 4;
            switch (countryFlag)
            {
                case 1:
                    country = GetFlagStr(ref offset, ref countryFlag, ref endIpOff);
                    offset = endIpOff + 8;
                    local = (1 == countryFlag) ? string.Empty : GetFlagStr(ref offset, ref countryFlag, ref endIpOff);
              
                    break;
                case 2:
                    country = GetFlagStr(ref offset, ref countryFlag, ref endIpOff);
                    offset = endIpOff + 8;
                    local = (1 == countryFlag) ? string.Empty : GetFlagStr(ref offset, ref countryFlag, ref endIpOff);
                    break;
                default:
                    country = GetFlagStr(ref offset, ref countryFlag, ref endIpOff);
                    local = GetFlagStr(ref offset, ref countryFlag, ref endIpOff);
                    break;
            }
            return country;
        }

        private string GetFlagStr(ref uint offset, ref int countryFlag, ref uint endIpOff)
        {
            int flag = 0;
            while (true)
            {
                flag = data[offset];
                //没有重定向
                if (flag != 1 && flag != 2)
                {
                    break;
                }
                if (flag == 2)
                {
                    countryFlag = 2;
                    endIpOff = offset - 4;
                }
                offset = (uint)data[1 + offset] + (((uint)data[2 + offset]) << 8) + (((uint)data[3 + offset]) << 16);
            }
            if (offset < 12)
            {
                return string.Empty;
            }
            return GetStr(ref offset);
        }

        /// <summary>
        /// 读取字符串...
        /// </summary>
        /// <param name="offset"></param>
        /// <returns></returns>
        private string GetStr(ref uint offset)
        {
            byte lowByte = 0;
            byte highByte = 0;
            StringBuilder stringBuilder = new StringBuilder(16);
            while (true)
            {
                lowByte = data[offset++];
                if (lowByte == 0)
                {
                    return stringBuilder.ToString();
                }
                if (lowByte > 0x7f)
                {
                    highByte = data[offset++];
                    if (highByte == 0)
                    {
                        return stringBuilder.ToString();
                    }
                    stringBuilder.Append(encoding.GetString(new byte[] { lowByte, highByte }));
                }
                else
                {
                    stringBuilder.Append((char)lowByte);
                }
            }
        }
    }
}
